core/src/main/java/org/apache/druid/storage/StorageConnector.java

```diff
@@ -75,6 +75,19 @@
    */
   InputStream read(String path) throws IOException;
 
+  /**
+   * Reads the data present for a given range at the path in the underlying storage system.
+   * Most implementations prepend the input path with a basePath.
+   * The caller should take care of closing the stream when done or in case of error. Further, the caller must ensure
+   * that the start offset and the size of the read are valid parameters for the given path for correct behavior.
+   * @param path The path to read data from
+   * @param from Start offset of the read in the path
+   * @param size Length of the read to be done
+   * @return InputStream starting from the given offset limited by the given size
+   * @throws IOException if the path is not present or the unable to read the data present on the path
+   */
+  InputStream readRange(String path, long from, long size) throws IOException;
```

Comment:
Should the behaviour of the implementations be defined here under following conditions:
1. from > size of underlying storage (undefined, error, or we return empty output)
2. from + size > size of underlying storage (undefined, error, or we chomp the stream)
3. from < 0  (undefined, error, or from indicates index from the end)
4. size < 0  (undefined, error, or we read in reverse)

Reply:
have added some validation checks for inputs in local and s3 connector. haven't done the size based check in s3 connector yet since it'll require a status api call for the object first - I'm not sure if it is necessary to do that check as of now.